internal/proxier/proxy.go

```diff
@@ -64,58 +75,257 @@ type ProxyOpts struct {
-			EventType: e.Event,
-			Service:   e.NewObject.(*corev1.Service),
+	if svc.DeletionTimestamp != nil {
+		p.pfrequest <- PortForwardRequest{
+			DeletePortForwardRequest: &DeletePortForwardRequest{
+				Service: ServiceInfo{Namespace: svc.Namespace, Name: svc.Name},
+			},
 		}
-	})
-	if err != nil {
-		return err
+		return nil
 	}
 
-	log.Info("waiting for kubernetes handlers to finish")
-	<-handlerDoneChan
+	existingForward := p.worker.portForwards[key]
+	if existingForward == nil {
+		//create a new port forward
+		p.createPortforward(svc, "")
+		return nil
+	}
 
-	log.Info("waiting for port-forward worker to finish")
-	<-pfdoneChan
+	e, exists, err := p.endpointsInformer.GetStore().GetByKey(key)
+	if !exists || err != nil {
+		// no endpoints for service nothing we can do atm
+		return nil
+	}
+	endpoints := e.(*corev1.Endpoints)
+
+	switch existingForward.Status {
+	case PortForwardStatusWaiting:
+		for _, subset := range endpoints.Subsets {
+			for _, address := range subset.Addresses {
+				if address.TargetRef != nil && address.TargetRef.Kind == PodKind {
+					p.createPortforward(svc, "endpoint became available")
+				}
+			}
+		}
+
+	case PortForwardStatusRunning:
+		if endpoints, err := p.endpointsInformer.GetIndexer().ByIndex("pods", existingForward.Pod.Name); err == nil && len(endpoints) == 0 {
+			p.createPortforward(svc, fmt.Sprintf("endpoints '%s' was removed", existingForward.Pod.Key()))
+		}
+	case PortForwardStatusRecreating:
+		//make exhaustive linter happy
+	}
 
 	return nil
 }
 
+func (p *Proxier) createPortforward(svc *corev1.Service, recreate string) {
+	info := ServiceInfo{Namespace: svc.Namespace, Name: svc.Name}
+	// resolve the service ports using endpoints if possible.
+	resolvedPorts, _, err := kube.ResolveServicePorts(svc)
+	if err != nil {
+		return
+	}
+
+	ports := make([]string, len(svc.Spec.Ports))
+	for i, p := range resolvedPorts {
+		ports[i] = fmt.Sprintf("%d:%d", p.Port, p.TargetPort.IntValue())
+	}
+	req := CreatePortForwardRequest{
+		Service: info,
+		Ports:   ports,
+		Hostnames: []string{
+			info.Name,
+			fmt.Sprintf("%s.%s", info.Name, info.Namespace),
+			fmt.Sprintf("%s.%s.svc", info.Name, info.Namespace),
+			fmt.Sprintf("%s.%s.svc.%s", info.Name, info.Namespace, p.opts.ClusterDomain),
+		},
+	}
+	// hack for basic support of statful sets.
+	// grab the first endpoint to build the name. This sucks, but it's
+	// needed for Outreach's usecases. Please remove this.
+	if obj, exists, err := p.endpointsInformer.GetStore().GetByKey(svc.Namespace + "/" + svc.Name); err == nil && exists {
+		endpoints := obj.(*corev1.Endpoints)
+		refName := ""
+	loop:
+		for _, sub := range endpoints.Subsets {
+			for _, a := range sub.Addresses {
+				if a.TargetRef != nil && a.TargetRef.Kind == PodKind {
+					refName = a.TargetRef.Name
+					break loop
+				}
+			}
+		}
+		if refName != "" {
+			name := fmt.Sprintf("%s.%s", refName, info.Name)
+			req.Hostnames = append(req.Hostnames,
```

Comment:
I think this needs to be updated to reflect the changes in #42

Reply:
Whats missing? I don't see it. I thought I incorporated the changes from #42.

Reply:
Ahhhh, I read this wrong. You did! Sorry!